bftengine/src/bftengine/SigManager.cpp

```diff
@@ -240,5 +229,16 @@ void SigManager::sign(const char* data, size_t dataLength, char* outSig, uint16_
 
 uint16_t SigManager::getMySigLength() const { return (uint16_t)mySigner_->signatureLength(); }
 
+void SigManager::setClientPublicKey(const std::string& key, PrincipalId id, KeyFormat format) {
+  if (replicasInfo_.isIdOfExternalClient(id)) {
+    std::unique_lock lock(mutex_);
+    verifiers_.insert_or_assign(id, std::make_shared<RSAVerifier>(key.c_str(), format));
```

Comment:
I don't that insert_or_assign should be used here, at least not like that. You are interested at this stage to replace keys right?
That means if a key was not in the map, something is wrong with the specific operation you want to perform.

I agree that we might want in the future to have an option to insert a new key, but the 2 operations should be distinct, at least in terms of catching errors and warning the user about errors.

For now, the only operation we want to perform is to "replace the key with an already existing one".
So if the key was not there and inserted, something very wrong happened, and the caller won't know about it.

I suggest adding some flag "replaceOnly" on the intention of the user and. according to that perform the operation.
If replaceOnly is true, only accept assign (throw an exception if a key wasn't there). else use insert_and_assign.

Reply:
`replicasInfo_.isIdOfExternalClient(id)` is checking the validity of the client, why to restrict ourselves? It can be a new client after reconfiguration for example. After all, SigManager is responsible for holding signer and verifiers by client id and all the key exchange logic is out of its scope.

Reply:
But if there is only one use case (see the other comment https://github.com/vmware/concord-bft/pull/1689#discussion_r672129719 ), then any other case is an error (a bug).
I didn't say to restrict. I suggested making the interface more clear, to reflect the user intention and catch bugs:
`void SigManager::setClientPublicKey(const std::string& key, PrincipalId id, KeyFormat format, boolkeyMustExist)`

if boolkeyMustExist is true then only insert.
if boolkeyMustExist is false,  insert_or_assign.

Reply:
Sorry, from my point of view if `KeyExchangeMgr` has decided that key exchange was legal, it just sets the key for the client. I don't see any case when I use the `boolKeyMustExist`.